driver-core/src/test/functional/com/mongodb/client/model/expressions/StringExpressionsFunctionalTest.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client.model.expressions;
+
+import org.junit.jupiter.api.Test;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+
+import static com.mongodb.client.model.expressions.Expressions.of;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+@SuppressWarnings({"ConstantConditions"})
+class StringExpressionsFunctionalTest extends AbstractExpressionsFunctionalTest {
+    // https://www.mongodb.com/docs/manual/reference/operator/aggregation/#string-expression-operators
+
+    private final String sushi = "\u5BFF\u53F8";
+    private final String jalapeno = "jalape\u00F1o";
+
+    @Test
+    public void literalsTest() {
+        assertExpression("", of(""), "''");
+        assertExpression("abc", of("abc"), "'abc'");
+        assertThrows(IllegalArgumentException.class, () -> of((String) null));
+        assertExpression(sushi, of(sushi), "'" + sushi + "'");
+    }
+
+    @Test
+    public void concatTest() {
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/concat/
+        assertExpression(
+                "abc".concat("de"),
+                of("abc").concat(of("de")),
+                "{'$concat': ['abc', 'de']}");
+    }
+
+    @Test
+    public void toLowerTest() {
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/toLower/
+        assertExpression(
+                "ABC".toLowerCase(),
+                of("ABC").toLower(),
+                "{'$toLower': 'ABC'}");
+    }
+
+    @Test
+    public void toUpperTest() {
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/toUpper/ (?)
+        assertExpression(
+                "abc".toUpperCase(),
+                of("abc").toUpper(),
+                "{'$toUpper': 'abc'}");
+    }
+
+    @Test
+    public void strLenTest() {
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/strLenCP/ (?)
+        // TODO naming: just strLen, lengthCP, lengthCodePoints, stringLengthInCodePoints?...
```

Comment:
> MQL substr method maps to substrBytes, but mine maps to substrCP

Yeah, I noticed that. I'd prefer `*CP` method names, so that we match MQL, and users will also be less confused. However, since one of the key ideas is to abstract the API from MQL, I can live with the current naming.

Reply:
Alright, I'll leave this for now but I think perhaps we should give all the method names a final review for consistency. Worth noting that, as you pointed out, strLen does not match Java's length method. See: https://github.com/mongodb/mongo-java-driver/pull/1032#discussion_r1021694487